### PR TITLE
Test builtin(front_facing)

### PR DESCRIPTION
**Pull request by greggman**
I'm not sure if I should test more sets of triangles

The current set is this

image

Where the red pixels are samples when sampleCount = 4, the white square is clip space (the 4x4 pixels of the texture) and the 2 yellow triangles are the triangles being drawn.

PS: The code to draw this diagram is [here](https://jsgist.org/?src=c31c34b617f65c7509d425a62e976337). I made it so pick vertex coordinates that wouldn't go through the center of samples. It might be useful to pick other triangles.

<hr>

**Requirements for PR author:**

- [X] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [X] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [X] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts [hunk 1]
```diff
@@ -869,7 +941,124 @@ g.test('inputs,sample_index')
...
+      nearFar,
+      sampleCount,
+      frontFace,
+      interpolation: { type, sampling },
+    } = t.params;
+    //
+    // We're drawing 2 triangles starting at y = -2 to y = +2
+    //
+    //  -1   0   1
+    //   +===+===+  2
```
Comment (shrekshao) [inline src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts:992]: The diagonal line should be flipped

File: src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts [hunk 2]
```diff
@@ -869,7 +941,124 @@ g.test('inputs,sample_index')
...
+    //   |   |/  |       |
+    //   +---+---+  0    | viewport
+    //   |  /|   |       |
+    //   +---+---+ -1  <---
+    //   |/  |   |
+    //   +===+===+ -2
+
+    // prettier-ignore
+    const clipSpacePoints = [
+      // cw
```
Comment (shrekshao) [inline src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts:1004]: Does the "cw" and "ccw" comment also need to flip here?

File: src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts [hunk 3]
```diff
@@ -403,6 +450,13 @@ function computeFragmentSampleIndex({ sampleIndex }: FragData) {
   return [sampleIndex, 0, 0, 0];
 }
 
+/**
+ * Computes 'builtin(front_facing)'
+ */
+function computeFragmentFrontFacing({ frontFacing }: FragData) {
+  return [frontFacing ? 0 : 1, 0, 0, 0];
```
Comment (shrekshao) [inline src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts:457]: corresponding `outputCode: 'vec4f(select(0.0, 1.0, fin.frontFacing), 0, 0, 0)'`

```suggestion
  return [frontFacing ? 1 : 0, 0, 0, 0];
```

**Review by shrekshao — COMMENTED** [review]
If I'm understanding correctly and it's passing, then there might be some double flip making it pass

File: src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts [hunk 4]
```diff
@@ -1,16 +1,16 @@
...
 * test builtin(position)
-* test interpolation
+* test @interpolate
+* test builtin(sample_index)
+* test builtin(front_facing)
 
-The current tests draw a single triangle with clip space coordinates [1, 1], [-3, 1], [1, -3].
-This means they render to all pixels in the textures. To fully test centroid interpolation
-probably requires drawing various triangles that only cover certain samples. That is TBD.
+Note: @interpolate settings and sample_index effect whether or not the fragment shader
```
[the remark this answers is not in the document]
Comment (greggman) [reply src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts:8]: done

File: src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts (hunk 1 above)
Comment (greggman) [reply src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts:992]: done

File: src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts (hunk 3 above)
Comment (greggman) [reply src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts:457]: done

File: src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts (hunk 2 above)
Comment (greggman) [reply src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts:1004]: done

**Comment by greggman** [comment]
> If I'm understanding correctly and it's passing, then there might be some double flip making it pass

I flipped `[frontFacing ? 1 : 0, ....` as you suggested and flipped `isTriangleClockwise`

File: src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts [hunk 5]
```diff
@@ -287,7 +287,7 @@ function isTriangleClockwise(windowPoints: readonly number[][]) {
     const p1 = windowPoints[(i + 1) % 3];
     sum += p0[kX] * p1[kY] - p1[kX] * p0[kY];
   }
-  return sum < 0;
+  return sum >= 0;
```
Comment (shrekshao) [inline src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts:290]: https://math.stackexchange.com/questions/1324179/how-to-tell-if-3-connected-points-are-connected-clockwise-or-counter-clockwise

Seems that clockwise is <0 though. Maybe there is something else that's flipped?

Comment (greggman) [reply src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts:290]: IIRC it's that in window coords  0,0 is at the top left in WebGPU vs OpenGL where it's bottom left, so triangles are flipped.

Comment (shrekshao) [reply src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts:290]: Got it.

**Review by shrekshao — APPROVED** [review]
LGTM

**Pull request merged**